CHANGELOG.md

```diff
@@ -1,7 +1,10 @@
 x.x.x Release notes (yyyy-MM-dd)
 =============================================================
+### Breaking Changes
+* On Node.js and Electron, change notification callbacks will no longer be invoked directly by the `Realm.write()` call. Instead, they will be invoked asynchronously on a subsequent event loop tick. 
```

Comment:
This seems like a significant regression in functionality. Invoking the change callbacks synchronously makes it difficult to accidentally read the updated data before the callback fires. If the change callback does anything important then any code which runs in between them is very likely to be incorrect.

Reply:
I wasn't able to figure out a v8 API that allows us to detect whether the change notification callback fires because of a call to `write()` or because another thread/process committed a write - that is to say `v8::Isolate::GetCallingContext()` which we called in `node::Function::callback` is removed. I also checked n-api and there isn't anything that seems useful there either.

Reply:
Do you have any ideas how we can implement this functionality differently?

Reply:
After digging into this more, the problem isn't that MakeCallback makes it async; it actually is still sync, but it runs microtasks before invoking the callback. This seems to be the correct thing to do when we're invoking a top-level callback (the tests hang in several places that look valid without it), but when we're invoking a callback from within a function called by JS it means that we process microtasks from within that function, which is definitely not something we want to do.

Newer versions of v8 don't seem to have anything comparable. GetCallingContext() walks the call stack looking for JS frames, but with tail calls there may not be any JS frames on the call stack even if we were called by JS.

I think maybe what we need is a node-specific version of `EventLoopSignal` that wraps the invocation of the callback in a `CallbackScope`. Doing something like the following would probably fix it:

```
{
    auto isolate = Isolate::GetCurrent();
    v8::HandleScope handle_scope(isolate);
    node::CallbackScope callback_scope(isolate);
    data.callback();
}
```

The problem with this is that it's a node dependency in object store, which maybe complicates things, and it processes microtasks more often than we currently do (not all event loop signals end up invoking a JS function), which may or may not matter.

An alternative would be to track if we're the top level of execution ourselves and do the current logic of maybe using MakeCallback based on that. A crude way of doing that would be to do something like:

```
// EventLoopSignal callback
{
    g_in_top_level_callback = true;
    data.callback();
    g_in_top_level_callback = false;
}

template<>
inline v8::Local<v8::Value> node::Function::callback(v8::Isolate* isolate, const v8::Local<v8::Function> &function, const v8::Local<v8::Object> &this_object, size_t argc, const v8::Local<v8::Value> arguments[]) {
    if (!g_in_top_level_callback) {
        // if there are any JavaScript frames on the stack below this one we don't need to
        // go through the trouble of calling MakeCallback. MakeCallback is only for when a
        // thread with no JavaScript frames on its stack needs to call into JavaScript, like in
        // an uv_async callback.
        return call(isolate, function, this_object, argc, arguments);
    }

    v8::TryCatch trycatch(isolate);
    
    auto recv = this_object.IsEmpty() ? isolate->GetCurrentContext()->Global() : this_object;
    g_in_top_level_callback = false;
    auto result = ::node::MakeCallback(isolate, recv, function, (int)argc, const_cast<v8::Local<v8::Value>*>(arguments));
    g_in_top_level_callback = true;

    if (trycatch.HasCaught()) {
        ::node::FatalException(isolate, trycatch);
    }
    return result;
}
```

Maybe there's a better way to do this.